[Feature] Add reduction parameter to Off-Policy losses.

albertbou92 opened the pull request:
## Description

Same as https://github.com/pytorch/rl/pull/1890, but for Off-Policy losses.

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

vmoens commented on test/test_cost.py:
```diff
@@ -481,7 +481,10 @@ def _create_seq_mock_data_dqn(
     @pytest.mark.parametrize("action_spec_type", ("one_hot", "categorical"))
     @pytest.mark.parametrize("td_est", list(ValueEstimators) + [None])
-    def test_dqn(self, delay_value, double_dqn, device, action_spec_type, td_est):
+    @pytest.mark.parametrize("reduction", [None, "none", "mean", "sum"])
```
I think at this stage we must be careful not to overparametrize the tests. Maybe we could just run a dedicated test for DQN that runs it on a simple unique scenario and checks that the loss values all have the right shape?

vmoens commented on test/test_cost.py:
```diff
@@ -688,8 +702,9 @@ def test_dqn_tensordict_run(self, action_spec_type, td_est):
         "action_spec_type", ("mult_one_hot", "one_hot", "categorical")
     )
     @pytest.mark.parametrize("td_est", list(ValueEstimators) + [None])
+    @pytest.mark.parametrize("reduction", [None, "none", "mean", "sum"])
```
ditto

vmoens commented on test/test_cost.py:
```diff
@@ -2086,6 +2120,7 @@ def _create_seq_mock_data_td3(
     @pytest.mark.parametrize("td_est", list(ValueEstimators) + [None])
     @pytest.mark.parametrize("use_action_spec", [True, False])
     @pytest.mark.parametrize("dropout", [0.0, 0.1])
+    @pytest.mark.parametrize("reduction", [None, "none", "mean", "sum"])
```
ditto

vmoens commented on test/test_cost.py:
```diff
@@ -2769,6 +2814,7 @@ def _create_seq_mock_data_sac(
     @pytest.mark.parametrize("num_qvalue", [1, 2, 4, 8])
     @pytest.mark.parametrize("device", get_default_devices())
     @pytest.mark.parametrize("td_est", list(ValueEstimators) + [None])
+    @pytest.mark.parametrize("reduction", [None, "none", "mean", "sum"])
```
ditto

vmoens commented on test/test_cost.py:
```diff
@@ -3568,6 +3624,7 @@ def _create_seq_mock_data_sac(
     @pytest.mark.parametrize("target_entropy_weight", [0.01, 0.5, 0.99])
     @pytest.mark.parametrize("target_entropy", ["auto", 1.0, 0.1, 0.0])
     @pytest.mark.parametrize("td_est", list(ValueEstimators) + [None])
+    @pytest.mark.parametrize("reduction", [None, "none", "mean", "sum"])
```
ditto

vmoens commented on test/test_cost.py:
```diff
@@ -4202,7 +4269,8 @@ def _create_seq_mock_data_redq(
     @pytest.mark.parametrize("device", get_default_devices())
     @pytest.mark.parametrize("td_est", list(ValueEstimators) + [None])
-    def test_redq(self, delay_qvalue, num_qvalue, device, td_est):
+    @pytest.mark.parametrize("reduction", [None, "none", "mean", "sum"])
```
ditto

vmoens reviewed (commented):
Thanks so much this is an amazing feature!

We should also do the deprecated REDQ because i'll de-deprecate it soon (it runs faster than the other lol)

albertbou92 commented:
I have changed the test so each loss has a dedicated test for the `reduction` parameter. Also added the `reduction` parameter to deprecated REDQ.

For on-policy losses we did not have dedicated tests for `reduction`, we added them to the general test (e.g. test_ppo()). Should we change that too? And if so, in this PR? maybe cleaner to do it in a separate PR.

vmoens commented:
Another PR seems appropriate! Thanks!
